koa/koa.d.ts

```diff
@@ -1,182 +1,654 @@
 // Type definitions for Koa 2.x
 // Project: http://koajs.com
-// Definitions by: DavidCai1993 <https://github.com/DavidCai1993>
+// Definitions by: DavidCai1993 <https://github.com/DavidCai1993>, jKey Lu <https://github.com/jkeylu>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 /* =================== USAGE ===================
 
     import * as Koa from "koa"
     const app = new Koa()
 
-    async function (ctx: Koa.Context, next: Function) {
-      // ...
-    }
+    app.use(async (ctx, next) => {
+        // ...
+    });
 
  =============================================== */
 /// <reference path="../cookies/cookies.d.ts"/>
 /// <reference path="../node/node.d.ts"/>
 
+
+declare namespace Koa {
+
+    // These open interfaces may be extended in an application-specific manner via declaration merging.
+    // See for example method-override.d.ts (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/method-override/method-override.d.ts)
+    export interface Request { }
+    export interface Response { }
+    export interface Context { }
+    export interface Application { }
+}
+
 declare module "koa" {
-    import { EventEmitter } from "events";
-    import * as cookies from "cookies";
-    import * as http from "http";
-    import * as net from "net";
-
-    namespace Koa {
-        export interface Context extends Request, Response {
-            app: Koa;
-            req: http.IncomingMessage;
-            res: http.ServerResponse;
-            request: Request;
-            response: Response;
+    import { EventEmitter } from 'events';
+    import * as cookies from 'cookies';
+    import * as http from 'http';
+    import * as net from 'net';
 
-            cookies: cookies.ICookies;
-            originalUrl: string;
-            state: any;
+    interface IRequest {
```

Comment:
This interface should be documented: it's one letter off from `Request`, and it's actually the supertype of `BaseRequest` -- what are the relationships between these types?
Also, we don't usually start interface names with `I`, although you'll of course need some way to distinguish this from `Request`.